docs/connery.man
.\" Man page for connery
.TH Connery 1 "2013-15-01" "Linux" "Connery IRC Bot"

.SH NAME
connery - A highly customizable Internet Relay Chat bot

.SH SYNOPSIS
.\" Options to actually run / quit
.B connery
.RB [ \-q | \--quit ]
.RB [ \-k | \--kill ]
.RB [ \-d | \--fork ]
.RB [ \--quiet ]
.RB [ -c
.IR filename
.RB |  \--config=
.IR filename ]
.\" Options to configure or get help
.br
.B connery
.RB [ \-h | \--help ]
.br
.B connery
.RB [ \-m | \--migrate ]
.br
.B connery
.RB [ \-w | \--configure-all ]
.RB [ \--configure-modules ]
.RB [ \--configure-database ]

.SH DESCRIPTION
.B connery
runs a bot that provides some usefull (and some useless) commands in one or more
Internet Relay Chat channels. It reads configuration options, such as the
network to connect to, channels to join, and more from a config file specified
by the 
.B -c
option. If no config file is specified,
.B ~/.connery/default.cfg
is used. If the configuration file being used does not exist, a brief
configuration wizard will be run to set up the basic options, and optionally set
up a database for user data and any modules that may need configuration.
.P
In most cases, simply running 
.B connery
will be sufficient; the bot will start after initial configuration, and will run
until it is told to quit through IRC.

.SH OPTIONS
.TP 5
\-h or --help
Show a brief summary of these options.
.TP
\-c
.IR filename
or --config=
.IR filename
.br
Run Connery with the configuration options from the specified file. An absolute
pathname can be given, or the name of a file within the configuration directory
may also be given.
.B ~/.connery
is the preferred location, but this and 
.B -c
will also look in 
.B ~/.jenni
and 
.B ~/.phenny
for legacy reasons. When the suffix of the file is
.B .cfg
, it may be omitted.
.TP
\-d or --fork
Run Connery in the background. The bot will safely run in the background, and can
later be killed with 
.B -q
or 
.B -d
\.
.TP
\-q or --quit
Gracefully quit a daemonized instance of Connery. This option determines the 
instance to quit by the name of the config file given by the 
.B -c
option, if used, so the same argument must be given with 
.B -c
when using 
.B -q.
.TP
\-k or --kill
Terminates a daemonized instance of Connery. This option determines the instance
to kill by the name of the config file given by the 
.B -c
option, if used, so the same argument must be given with 
.B -c
when using 
.B -k.
.TP
\-l or --list
List all config files found in the usual configuration file directories:
.B ~/.connery ~/.jenni
and 
.B ~/.phenny
.TP
\-m or --migrate
Attempt to migrate a configuration file for an older version of the bot to the 
new format. Versions prior to 3.1 used a configuration file written in Python,
which can be converted with this tool.
.TP
\--quiet
Suppress all output from the bot to the terminal.
.TP
\-w or --configure-all
Run the initial configuration wizard. This can be run with
.B -c
to configure a specified file, or without it to work on the default file.
.TP
\--configure-modules
Check for modules with configuration options, and allow the user to modify those
options. Changes will be added to the default file, or the file given to
.B -c
if used.
.TP
\--configure-database
Run a wizard for configuring the user information database options. Changes
will be added to the default file, or the file given to
.B -c
if used.


NEWS
Changes between 5.3.0 and 5.4.0
===============================
* Sean Connery, mother fucker.

Changes between 5.2.0 and 5.3.0
===============================
Module changes (for users):
* The YouTube module is removed due to breakage of its API by Google
* Fixes for unicode in channel logging
* .py hits an updated service running Python 2.7, rather than 2.5
* Multiple new features are added in the reddit module
* Wind direction arrows in weather now point the right way

API changes (for developers):
* Time and calculation tools are split out into multiple modules. The moved things will also be available in tools itself until 6.0, whereupon they will only be available in their new locations.
* Functions and classes used internally for scheduling jobs are moved. They are still available in their old location, but are no longer documented and will not be considered part of the public API in 6.0.


Changes between 5.1.1 and 5.2.0
===============================
Module changes (for users):
* An exception that failed the chanlogs module from loading is fixed
* Meetbot meeting subjects no longer fail with unicode

Core changes (for users):
* The various ways of authenticating the bot are now configured using the same set of attributes. The old settings will be used as fallback until 6.0.0

API changes (for developers):
* Privilege decorators no longer cause an error when no message is given
* A new @intent decorator is added as a shortcut for triggering on a message with an intent

Changes between 5.0.0 and 5.1.1
===============================
Module changes (for users):
* Fixed a regression that caused getting weather for a nick to fail
* Bugs related to channel log filenames are fixed
* Channel logs can now use the bot's preferred time zone
* Getter and setter methods for timezone and format are more consistently named
* .seen persists across bot restarts
* .seen no longer shows message or channel unless used within the same channel
* Special characters in Wikipedia URLs are handled correctly

Core changes (for users):
* help_prefix can now be given in [core] to change the command prefix used in help

API changes (for developers):
* The Trigger object is now immutable, as expected
* New decorators for checking privileges on callables are added

Changes between 4.6.2 and 5.0.0
===============================
This release contains backwards-incompatible changes. See
http://connery.dftba.net/connery_5.html for more information.

Module changes (for users):
* YouTube no longer shows bizarre lengths when the bot is running under Python 3
* When Wikipedia links are posted, a snippet of the article is shown

Core changes (for users):
* ConneryDB is entirely rewritten, meaning migration will be needed to access old data
* Logging output to the debug channel is improved
* The name of the NickServ user can be configured with nickserv_name in [core]
* SSL is disabled on Python 2.7 when backports.ssl_match_hostname is not installed

API changes (for developers):
* Deprecated ConneryDB functions are removed
* CTCP actions are stripped prior to matching, and added to the 'intent' key of trigger.tags
* Deprecated Trigger functions are removed
* bot.debug is removed, in favor of standard Python logging
* tools.Nick is removed, in favor of the tools.Identifier introduced in 4.6.0

Changes between 4.6.1 and 4.6.2
===============================
This release starts preparations for Connery 5. See http://connery.dftba.net/connery_5.html
for more information.

Module changes (for users):
* Due to API deprecation and instability, .g now uses DuckDuckGo instead of Google

Core changes (for users):
* Fix remaining regression in db update function

Changes between 4.6.0 and 4.6.1
===============================
This release starts preparations for Connery 5. See http://connery.dftba.net/connery_5.html
for more information.

Module changes (for users):
* Fix regression in table creations and erroneously changed column names

Core changes (for users):
* Fix regression in db update function

API changes (for developers):
* Correctly print out deprecation warnings

Changes between 4.5.1 and 4.6.0
===============================
This release starts preparations for Connery 5. See http://connery.dftba.net/connery_5.html
for more information.

Module Changes (for users):
* The ^ operator in .calc is now equivalent to **
* .dice, .tr, and .addtrace give meaningful errors when no argument is given
* A number of database issues were fixed (more will be fixed in Connery 5)
* A number of Python 3-related issues were fixed.
* Malicious URLs are now detected with VirusTotal and malwaredomains
* .weather no longer shows pressure

Core changes (for users):
* The config wizard no longer configures the database, instead automatically creating a sqlite db
* MySQL and Postgres databases are deprecated; only sqlite will be supported in Connery 5
* Logging channel and level can be configured with logging_channel and logging_level in the config
* A --version switch was added to the connery command, to show the version in use
* Corrupt PID files will no longer prevent the bot from starting
* Non-ASCII values no longer crash the config wizard

API changes (for developers):
* Numerous functions were added to the db to ease the transition to Connery 5
* tools.Nick has been renamed tools.Identifier. Nick will be removed in Connery 5
* connery.web functions now specify a meaningful default user agent
* web.post now supports dicts as the payload, and can optionally return headers
* config.get was added as an alias to config.parser.get
* Callables can now handle more than one event by stacking decorators

Changes between 4.5.0 and 4.5.1
===============================
Module Changes (for users):
* Version checking module no longer causes an error at startup
* URL handling modules no longer repeat themselves after being reloaded
* .isup now adds a .com TLD if none is used
* IPython integration now works with older versions of IPython


connery.py
#!/usr/bin/env python2.7
# coding=utf8
"""
Connery - An IRC Bot
Copyright 2008, Sean B. Palmer, inamidst.com
Copyright © 2012-2014, Elad Alfassa <elad@fedoraproject.org>
Licensed under the Eiffel Forum License 2.

http://connery.dftba.net
"""
from __future__ import unicode_literals
from __future__ import print_function

import sys
from connery.tools import stderr

if sys.version_info < (2, 7):
    stderr('Error: Requires Python 2.7 or later. Try python2.7 connery')
    sys.exit(1)
if sys.version_info.major == 3 and sys.version_info.minor < 3:
    stderr('Error: When running on Python 3, Python 3.3 is required.')
    sys.exit(1)

import os
import argparse
import signal

from connery.__init__ import run, __version__
from connery.config import Config, create_config, ConfigurationError, wizard
import connery.tools as tools
import connery.web

homedir = os.path.join(os.path.expanduser('~'), '.connery')


def enumerate_configs(extension='.cfg'):
    configfiles = []
    if os.path.isdir(homedir):
        connery_dotdirfiles = os.listdir(homedir)  # Preferred
        for item in connery_dotdirfiles:
            if item.endswith(extension):
                configfiles.append(item)

    return configfiles


def find_config(name, extension='.cfg'):
    if os.path.isfile(name):
        return name
    configs = enumerate_configs(extension)
    if name in configs or name + extension in configs:
        if name + extension in configs:
            name = name + extension

    return os.path.join(homedir, name)


def main(argv=None):
    global homedir
    # Step One: Parse The Command Line
    try:
        parser = argparse.ArgumentParser(description='Connery IRC Bot',
                                         usage='%(prog)s [options]')
        parser.add_argument('-c', '--config', metavar='filename',
                            help='use a specific configuration file')
        parser.add_argument("-d", '--fork', action="store_true",
                            dest="deamonize", help="Deamonize connery")
        parser.add_argument("-q", '--quit', action="store_true", dest="quit",
                            help="Gracefully quit Connery")
        parser.add_argument("-k", '--kill', action="store_true", dest="kill",
                            help="Kill Connery")
        parser.add_argument('--exit-on-error', action="store_true",
                            dest="exit_on_error", help=(
                                "Exit immediately on every error instead of "
                                "trying to recover"))
        parser.add_argument("-l", '--list', action="store_true",
                            dest="list_configs",
                            help="List all config files found")
        parser.add_argument("-m", '--migrate', action="store_true",
                            dest="migrate_configs",
                            help="Migrate config files to the new format")
        parser.add_argument('--quiet', action="store_true", dest="quiet",
                            help="Supress all output")
        parser.add_argument('-w', '--configure-all', action='store_true',
                            dest='wizard', help='Run the configuration wizard.')
        parser.add_argument('--configure-modules', action='store_true',
                            dest='mod_wizard', help=(
                                'Run the configuration wizard, but only for the '
                                'module configuration options.'))
        parser.add_argument('-v', '--version', action="store_true",
                            dest="version", help="Show version number and exit")
        opts = parser.parse_args()

        # Step Two: "Do not run as root" checks.
        try:
            # Linux/Mac
            if os.getuid() == 0 or os.geteuid() == 0:
                stderr('Error: Do not run Connery with root privileges.')
                sys.exit(1)
        except AttributeError:
            # Windows
            if os.environ.get("USERNAME") == "Administrator":
                stderr('Error: Do not run Connery as Administrator.')
                sys.exit(1)

        if opts.version:
            py_ver = '%s.%s.%s' % (sys.version_info.major,
                                   sys.version_info.minor,
                                   sys.version_info.micro)
            print('Connery %s (running on python %s)' % (__version__, py_ver))
            print('http://connery.dftba.net/')
            return
        elif opts.wizard:
            wizard('all', opts.config)
            return
        elif opts.mod_wizard:
            wizard('mod', opts.config)
            return

        if opts.list_configs:
            configs = enumerate_configs()
            print('Config files in ~/.connery:')
            if len(configs) is 0:
                print('\tNone found')
            else:
                for config in configs:
                    print('\t%s' % config)
            print('-------------------------')
            return

        config_name = opts.config or 'default'

        configpath = find_config(config_name)
        if not os.path.isfile(configpath):
            print("Welcome to Connery!\nI can't seem to find the configuration file, so let's generate it!\n")
            if not configpath.endswith('.cfg'):
                configpath = configpath + '.cfg'
            create_config(configpath)
            configpath = find_config(config_name)
        try:
            config_module = Config(configpath)
        except ConfigurationError as e:
            stderr(e)
            sys.exit(2)

        if config_module.core.not_configured:
            stderr('Bot is not configured, can\'t start')
            # exit with code 2 to prevent auto restart on fail by systemd
            sys.exit(2)

        if not config_module.has_option('core', 'homedir'):
            config_module.dotdir = homedir
            config_module.homedir = homedir
        else:
            homedir = config_module.core.homedir
            config_module.dotdir = config_module.core.homedir

        if not config_module.core.logdir:
            config_module.core.logdir = os.path.join(homedir, 'logs')
        logfile = os.path.os.path.join(config_module.logdir, 'stdio.log')
        if not os.path.isdir(config_module.logdir):
            os.mkdir(config_module.logdir)

        config_module.exit_on_error = opts.exit_on_error
        config_module._is_deamonized = opts.deamonize

        sys.stderr = tools.OutputRedirect(logfile, True, opts.quiet)
        sys.stdout = tools.OutputRedirect(logfile, False, opts.quiet)

        # Handle --quit, --kill and saving the PID to file
        pid_dir = config_module.core.pid_dir or homedir
        if opts.config is None:
            pid_file_path = os.path.join(pid_dir, 'connery.pid')
        else:
            basename = os.path.basename(opts.config)
            if basename.endswith('.cfg'):
                basename = basename[:-4]
            pid_file_path = os.path.join(pid_dir, 'connery-%s.pid' % basename)
        if os.path.isfile(pid_file_path):
            with open(pid_file_path, 'r') as pid_file:
                try:
                    old_pid = int(pid_file.read())
                except ValueError:
                    old_pid = None
            if old_pid is not None and tools.check_pid(old_pid):
                if not opts.quit and not opts.kill:
                    stderr('There\'s already a Connery instance running with this config file')
                    stderr('Try using the --quit or the --kill options')
                    sys.exit(1)
                elif opts.kill:
                    stderr('Killing the connery')
                    os.kill(old_pid, signal.SIGKILL)
                    sys.exit(0)
                elif opts.quit:
                    stderr('Signaling Connery to stop gracefully')
                    if hasattr(signal, 'SIGUSR1'):
                        os.kill(old_pid, signal.SIGUSR1)
                    else:
                        os.kill(old_pid, signal.SIGTERM)
                    sys.exit(0)
            elif old_pid is None or (not tools.check_pid(old_pid)
                                     and (opts.kill or opts.quit)):
                stderr('Connery is not running!')
                sys.exit(1)
        elif opts.quit or opts.kill:
            stderr('Connery is not running!')
            sys.exit(1)
        if opts.deamonize:
            child_pid = os.fork()
            if child_pid is not 0:
                sys.exit()
        with open(pid_file_path, 'w') as pid_file:
            pid_file.write(str(os.getpid()))
        config_module.pid_file_path = pid_file_path

        # Step Five: Initialise And Run connery
        run(config_module)
    except KeyboardInterrupt:
        print("\n\nInterrupted")
        os._exit(1)
if __name__ == '__main__':
    main()
